### ISSUE TITLE
Error using TileDao  queryForTile(long column, long row, long zoomLevel) method

### ISSUE THREAD

**Issue by syleniaa** [ngageoint/geopackage-android#46]
I'm trying to use queryForTile(long column, long row, long zoomLevel) to retrive an specific map tile but it gives me an error. I'm only using the method, in fact the exception shows that the error it's not in my code, it's internal.

GeoPackage geoPackage = geoPackageTileWriter.getGeopackage();
TileDao tileDao = geoPackage.getTileDao(tileSource.name());

    if(tileDao != null) {
                .....
                TileRow tileRow = tileDao.queryForTile(col,row,zoom);
     }

#### Version Information:
  * GeoPackage Android Version: 2.0.0
  * GeoPackage Android Source: (e.g. Central Repository / Bintray, Release, Source Code Build)
  * Device or Emulator: Pixel XL
  * Android Studio Version: 3.0
  * Android Version Number: 7
  * Android SDK Version: 24
  * Platform & OS: Windows
  * Other Relevant Libraries: osmdroid

#### Expected Results:

TileRow tileRow = tileDao.queryForTile(col,row,(long)zoom);
I want to receive the specific tile row

#### Observed Results:
It throws an exception

#### Output:

java.lang.ClassCastException: android.database.sqlite.SQLiteCursor cannot be cast to mil.nga.geopackage.user.UserCursor
                                                                          at mil.nga.geopackage.user.UserConnection.convertCursor(UserConnection.java:171)
                                                                          at mil.nga.geopackage.user.UserConnection.handleCursor(UserConnection.java:158)
                                                                          at mil.nga.geopackage.user.UserConnection.query(UserConnection.java:145)
                                                                          at mil.nga.geopackage.user.UserConnection.query(UserConnection.java:53)
                                                                          at mil.nga.geopackage.user.UserConnection.query(UserConnection.java:18)
                                                                          at mil.nga.geopackage.user.UserCoreDao.queryForFieldValues(UserCoreDao.java:245)
                                                                          at mil.nga.geopackage.tiles.user.TileDao.queryForTile(TileDao.java:258)
                                                                          at cu.citi.idesoi.pcidroid.tilesprovider.modules.GeoPackageMapTileModuleProvider$TileLoader.loadTile(GeoPackageMapTileModuleProvider.java:179)
                                                                          at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase$TileLoader.run(MapTileModuleProviderBase.java:297)
                                                                          at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
                                                                          at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
                                                                          at java.lang.Thread.run(Thread.java:761)

**Comment by bosborn** [comment]
I was unable to immediately recreate your error.  The following test successfully runs when checking out 2.0.0.

```
public void testQueryForTile() throws SQLException {

	for(String table: geoPackage.getTileTables()){

		TileDao tileDao = geoPackage.getTileDao(table);

		for(TileMatrix tileMatrix: tileDao.getTileMatrices()){

			long zoomLevel = tileMatrix.getZoomLevel();
			long rows = tileMatrix.getMatrixHeight();
			long columns = tileMatrix.getMatrixWidth();

			for(long row = 0; row < rows; row++){
				for(long column  = 0; column < columns; column++){

					TileRow tileRow = tileDao.queryForTile(column, row, zoomLevel);
					TestCase.assertNotNull(tileRow);

				}
			}

		}

	}

	geoPackage.close();
}
```

Looking at the error, it looks like something is out of place with the cursor factory.  The [GeoPackageManager open call](https://github.com/ngageoint/geopackage-android/blob/f3ce5dc2f660c51614809b7ad1cddb0bc5376c96/geopackage-sdk/src/main/java/mil/nga/geopackage/GeoPackageManager.java#L398) passes a [GeoPackageCursorFactory](https://github.com/ngageoint/geopackage-android/blob/master/geopackage-sdk/src/main/java/mil/nga/geopackage/factory/GeoPackageCursorFactory.java) in when creating the [SQLiteDatabase](https://developer.android.com/reference/android/database/sqlite/SQLiteDatabase.html).  When retrieving a TileDao, a [TileCursor wrapper is registered](https://github.com/ngageoint/geopackage-android/blob/f3ce5dc2f660c51614809b7ad1cddb0bc5376c96/geopackage-sdk/src/main/java/mil/nga/geopackage/factory/GeoPackageImpl.java#L213).  What are the steps used to create/import and open the GeoPackage?  Are you using the GeoPackageFactory and GeoPackageManager?  If you are able to duplicate the error with a small example that could help as well.

**Comment by syleniaa** [comment]
Hi, thanks for answering.
I'm using the geopackage-android library with osmdroid library. I'm
developing an extension for using a geopackage file as cache. I create the
geopackage file in this class using the GeoPackageFactory and
GeoPackageManager like this

GeoPackageFactory.getManager(context).createFile(DATABASE_FILENAME,
geopackage_file)

and for opening the geopackage i use this

geopackage =  GeoPackageFactory.getManager(context).open(DATABASE_FILENAME);

after i fill my geopackage , later in my code when i try to retrieve an
specif tile it's when the error happens

long col = (long) pTile.getX();
long row = (long) pTile.getY();
long zoom = (long) pTile.getZoomLevel();

TileRow tileRow = tileDao.queryForTile(col,row,(long)zoom);

when i debug my code the error occurs here in the class UserConnection

protected TResult convertCursor(Cursor cursor) {
    return (TResult) cursor;
}

I'm trying to duplicate the error in a simple example

2018-02-21 23:09 GMT+01:00 Brian Osborn <notifications@github.com>:

**Comment by syleniaa** [comment]
I solved the problem in my app using 
Cursor cursor = geoPackage.rawQuery("SELECT tile_data " + "FROM \"" + tileSource.name() +"\" " + "WHERE tile_row=? AND tile_column=? AND zoom_level=?",
                        new String[]{String.valueOf(row), String.valueOf(col), String.valueOf(zoom)}); 
instead of 
tileDao.query... none of them worked for me, it kept throwing the same exception

**Comment by bosborn** [comment]
Yeah it's not ideal to have to work around in that manner.  osmdroid depends on geopackage-android, are you using the transitive dependency or depending on it explicitly?  I'm assuming your extension is either private or closed source at the moment?

Still seems like an issue with the cursor factory not being invoked to wrap the cursor as a TileCursor.

osmdroid uses FeatureCursor [here](https://github.com/osmdroid/osmdroid/blob/fe91f37af47f0055a4fa7f431eb68b9e5f8049ea/OpenStreetMapViewer/src/main/java/org/osmdroid/samplefragments/geopackage/GeopackageFeatures.java#L204).  I assume without any issues and it works similar to TileCursor.  Have you tried reading a feature from the FeatureDao to see if that cursor gets wrapped?

osmdroid also uses creates a TileDao [here](https://github.com/osmdroid/osmdroid/blob/fe91f37af47f0055a4fa7f431eb68b9e5f8049ea/osmdroid-geopackage/src/main/java/org/osmdroid/gpkg/tiles/raster/GeoPackageMapTileModuleProvider.java#L83) and passes it to a GeoPackageTileRetriever.  Under the covers this retrieves a TileCursor[ in TileCreator](https://github.com/ngageoint/geopackage-android/blob/f3ce5dc2f660c51614809b7ad1cddb0bc5376c96/geopackage-sdk/src/main/java/mil/nga/geopackage/tiles/retriever/TileCreator.java#L494).  Again, I assume without any issues.

Can you try something along these lines to see if a simple cursor factory test works?
```
final TileTable tileTable = tileDao.getTable();
SQLiteDatabase.CursorFactory cursorFactory = new SQLiteDatabase.CursorFactory(){
	public Cursor newCursor(SQLiteDatabase db, SQLiteCursorDriver driver, String editTable, SQLiteQuery query) {
		return new TileCursor(tileTable, new SQLiteCursor(driver, editTable, query));
	}
};
SQLiteDatabase sqlite = activity.openOrCreateDatabase(geoPackage.getName(), Context.MODE_PRIVATE, cursorFactory);

Cursor cursor = sqlite.rawQuery("SELECT * FROM " + tileDao.getTableName(), null);
TestCase.assertTrue(cursor.getCount() > 0);
TestCase.assertTrue(cursor instanceof TileCursor);

TileCursor tileCursor = (TileCursor) cursor;
while(tileCursor.moveToNext()){
	TileRow tileRow = tileCursor.getRow();
	TestCase.assertNotNull(tileRow);
}
tileCursor.close();
```

@spyhunter99 do you have any thoughts on what might be causing this?

**Comment by spyhunter99** [comment]
Hmm not at the moment. I haven't tested the code since the last update which i think was before the updates for extra large table columns. The last release of osmdroid, which was last week, was v6.0.1 and the module in question uses nga packages v2.0.0.

I just reran the basic test that we have it is crashing. We've turned on strict mode to ensure that we close all resource streams. 

````
02-22 18:41:48.989  1537  1545 E StrictMode: A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
02-22 18:41:48.989  1537  1545 E StrictMode: java.lang.Throwable: Explicit termination method 'close' not called
02-22 18:41:48.989  1537  1545 E StrictMode:    at dalvik.system.CloseGuard.open(CloseGuard.java:180)
02-22 18:41:48.989  1537  1545 E StrictMode:    at java.io.FileInputStream.<init>(FileInputStream.java:78)
02-22 18:41:48.989  1537  1545 E StrictMode:    at mil.nga.geopackage.factory.GeoPackageManagerImpl.importGeoPackage(GeoPackageManagerImpl.java:657)
02-22 18:41:48.989  1537  1545 E StrictMode:    at mil.nga.geopackage.factory.GeoPackageManagerImpl.importGeoPackage(GeoPackageManagerImpl.java:572)
02-22 18:41:48.989  1537  1545 E StrictMode:    at org.osmdroid.gpkg.tiles.raster.GeoPackageMapTileModuleProvider.<init>(GeoPackageMapTileModuleProvider.java:60)
02-22 18:41:48.989  1537  1545 E StrictMode:    at org.osmdroid.gpkg.tiles.raster.GeoPackageProvider.<init>(GeoPackageProvider.java:74)
02-22 18:41:48.989  1537  1545 E StrictMode:    at org.osmdroid.gpkg.tiles.raster.GeoPackageProvider.<init>(GeoPackageProvider.java:43)
02-22 18:41:48.989  1537  1545 E StrictMode:    at org.osmdroid.samplefragments.tileproviders.OfflinePickerSample.setProviderConfig(OfflinePickerSample.java:211)
02-22 18:41:48.989  1537  1545 E StrictMode:    at org.osmdroid.samplefragments.tileproviders.OfflinePickerSample.access$000(OfflinePickerSample.java:61)
02-22 18:41:48.989  1537  1545 E StrictMode:    at org.osmdroid.samplefragments.tileproviders.OfflinePickerSample$1.onSelectedFilePaths(OfflinePickerSample.java:131)
````

we set strict mode in the application class using the following
````
 if (BuildConfig.DEBUG && Build.VERSION.SDK_INT >=11) {
            StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
                .detectDiskReads()
                .detectDiskWrites()
                .detectNetwork()   // or .detectAll() for all detectable problems
                .penaltyLog()
                .build());
            StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
                .detectLeakedSqlLiteObjects()
                .detectLeakedClosableObjects()
                .penaltyLog()
                .penaltyDeath()
                .build());
        }
````

I'd highly recommend it at development time.

I've also ran into some bounding box issues, osmdroid v6 has tighter controls on the web mercator bounding box limits. All of my sample gpkgs are at exactly +-90 latitude, which is larger the max bounds of web mercator. I'm working a few changes to fix this.  Besides this, the ERDC whitehorse set is returning the following bounds
`N:85.05112877980659; E:-1.5028131257091932E7; S:8570731.107560238; W:-1.5067267015573943E7`
obviously the south part is wrong and is probably in your neck of the woods. Looks like it's off by 1E6

I'm still investigating the root cause of this ticket but I suspect it's a classpath issue.

**Comment by spyhunter99** [comment]
I've also seem similar values with east and west values, is it not lat/lon dd?

**Comment by bosborn** [comment]
Thanks for the suggestion.  Added the leaked sqlite and closeable objects strict mode to the tests and fixed a few issues & tests.  The issue you saw should only occur if an import GeoPackage exception occurs (GeoPackage already exists w/o override flag on, etc).  For the happy path the stream was being closed.  [Fixed it](https://github.com/ngageoint/geopackage-android/commit/a6d9dc3f537ce6a98de1fc999fd396b7444b9f24) to close regardless of successful outcome which will go out on next release.

The newest WhiteHorse GeoPackage I just downloaded has 3857 (Web Mercator) bounds of:
`N:8609866.866042253; E:-1.5028131257091932E7; S:8570731.107560243; W:-1.5067267015573943E7`

Is the code assuming the projection to be 4326 (WGS84)? My guess is the code is setting north to be the `min(8609866.866042253, 85.05112877980659)`.  It's probably setting the south to be the `max(8570731.107560243, -85.05112877980659)` which lucks out and doesn't invalidate the south coordinate.

We do [something similar in TileBoundingBoxUtils ](https://github.com/ngageoint/geopackage-core-java/blob/00d11324ec818dcf7d72e5582711ce47b1431d71/src/main/java/mil/nga/geopackage/tiles/TileBoundingBoxUtils.java#L1103)but the calling code must verify the bounding box is in degrees.

If you want to get the tile bounds in WGS84, something like this:
```
ProjectionTransform transform = tileDao.getProjection().getTransformation(ProjectionConstants.EPSG_WORLD_GEODETIC_SYSTEM);
BoundingBox boundingBox = transform.transform(tileDao.getBoundingBox());
```
